### PR TITLE
Boosted Dendrites Module and Experiments

### PR DESCRIPTION
**Pull request by mvacaporale**
This introduces the `BoostedDendritesBase` module which applies boosted dendritic activations (see docstring for example usage). Accompanying functionality and changes includes:
- Mixin `UpdateDendriteBoostStrength`
- Refactoring of `apply_dendrite` functions to accept an option indices arguments.
- A new metric plotting function `plot_winning_segment_distributions` 
- Boosted-dendritic networks and associated experiment configs.

Experiments can be found in `projects/meta_cl/experiments/dendrite_variants.py`. Initial results proved less effective than ANML. However, slower lower rates in the dendrites may make a difference as the accuracy is improved when the weights of the boosted dendritic network (particularly the modulated part) are completely frozen.

### CODE REVIEW

**Comment by mvacaporale** [comment]
Note: There's a lot here, but the main thing to review is `frameworks/dendrites/dendrite_segments_booster.py`, `frameworks/dendrites/functional/apply_dendrites.py`, and `frameworks/dendrites/metrics/plotting.py `. 

@karangrewal  Moving forward, I'm hoping the `BoostedDendrites` module is relatively straightforward for you to try and use in your experiments as you see fit.

**Comment by mvacaporale** [comment]
Note: There's a lot here, but the main thing to review is `frameworks/dendrites/modules/boosted_dendrites.py`, `frameworks/dendrites/functional/apply_dendrites.py`, and `frameworks/dendrites/metrics/plotting.py `. 

@karangrewal  Moving forward, I'm hoping the `BoostedDendrites` module is relatively straightforward for you to try and use in your experiments as you see fit. It works very similarly to the other apply-dendrite modules (e.g. `DendriticGate1d`) except that it automatically updates the duty-cycles and boost activations accordingly before choosing the winning segments. As well, you must be sure to update the boost strength every epoch (for instance, with the `UpdateDendriteBoostStrength` mixin)

**Review by lucasosouza — APPROVED** [review]
Looks good, I have nothing to add

File: nupic/research/frameworks/dendrites/modules/boosted_dendrites.py [hunk 1]
```diff
@@ -0,0 +1,195 @@
...
+
+    def boost_activations(self, dendrite_activations):
+        """
+        Boost dendrite activations according to the duty cycles of the winning segments.
+
+        :param dendrite_activations: output from a DendriteSegments module;
+                                     shape batch_size x num_units x num_segments
+        """
+        boosted = dendrite_activations.detach()
+        boosted *= torch.exp(-self.boost_strength * self.duty_cycles)
```
Comment (karangrewal) [inline nupic/research/frameworks/dendrites/modules/boosted_dendrites.py:123]: (This relates to the comment above) `boosted` will approach 0 as `boost_strength` approaches infinity. I find that to be unintuitive.

**Review by karangrewal — APPROVED** [review]
Please address my comments before merging. Otherwise, the code looks great.

File: nupic/research/frameworks/dendrites/metrics/plotting.py [hunk 2]
```diff
@@ -337,3 +338,73 @@ def plot_entropy_distribution(winning_mask, targets):
...
+    # Generate subplots.
+    fig, axs = plt.subplots(1, num_units_to_plot, figsize=(6 * num_units_to_plot, 4))
+    if num_units_to_plot == 1:
+        axs = [axs]  # ensure this is subscriptable
+
+    # Generate a plot for each unit.
+    num_segments = winning_mask.shape[2]
+    for i, unit in enumerate(units):
+        indices = winning_indices[:, unit].cpu().numpy()
+        plot_winning_segment_distribution(indices, num_segments, unit=unit, ax=axs[i])
```
[the remark this answers is not in the document]
Comment (mvacaporale) [reply nupic/research/frameworks/dendrites/metrics/plotting.py:380]: Good point, fixed.

File: nupic/research/frameworks/dendrites/metrics/plotting.py [hunk 3]
```diff
@@ -337,3 +338,73 @@ def plot_entropy_distribution(winning_mask, targets):
...
+    seed=0
+):
+    """
+    Plot the distribution of winning segments for the list of units (defaults to just
+    the first):
+
+    :param dendrite_activations_: unused
+    :param winning_mask: the winning mask of segments;
+                         shape num_samples x num_units x num_segments
+    :param targets_: unused
```
[the remark this answers is not in the document]
Comment (mvacaporale) [reply nupic/research/frameworks/dendrites/metrics/plotting.py:357]: Good point, fixed.

File: nupic/research/frameworks/dendrites/modules/boosted_dendrites.py [hunk 4]
```diff
@@ -0,0 +1,195 @@
...
+
+    # Update boost_strength after each epoch.
+    segments_booster.update_boost_strength()
+    ```
+
+    :param num_units: number of units i.e. neurons;
+                      each unit has it's own set of dendrite segments
+    :param num_segments: number of dendrite segments per unit
+    :param boost_strength: initial boost strength; the higher the value, the more
+                           scarcely winning segments will be promoted
```
[the remark this answers is not in the document]
Comment (mvacaporale) [reply nupic/research/frameworks/dendrites/modules/boosted_dendrites.py:71]: I think my wording is confusing. The intuition and math are meant to be identical to the kwinners boosting functionality. It should read "segments that rarely win will be promoted, and the high the boost strength, the greater this effect". Is that more clear?

File: nupic/research/frameworks/dendrites/modules/boosted_dendrites.py (hunk 1 above)
Comment (mvacaporale) [reply nupic/research/frameworks/dendrites/modules/boosted_dendrites.py:123]: When `boost_strength` is 0, then `boosted` will be 1 (exp(0) = 1). Otherwise, boosted will be somewhere between 0 and 1. A higher duty cycle will make boosted closer to 0 than a lower duty cycle. For instance,
```
boost_strength = 0.3
np.exp(-boost_strength * 4) # lower duty cycle
>>> 0.30
np.exp(-boost_strength * 5) # lower duty cycle
>>> 0.22
```

Thus the units with the lower duty cycle have greater multipliers applied to their activations. This effect is amplified for higher boost strengths earlier on in training. Say `boost_strength = 1`, then in the example from before
```
boost_strength = 1
np.exp(-boost_strength * 4)
>>> 0.018
np.exp(-boost_strength * 5)
>>> 0.006
```
Now, the high duty cycle unit must have an activation three times greater than the lower duty cycle to win. Whereas before (`boost_strength = 0.3`) it only needed to be about 50% higher to win.

File: nupic/research/frameworks/dendrites/modules/boosted_dendrites.py (hunk 4 above)
[the remark this answers is not in the document]
Comment (mvacaporale) [reply nupic/research/frameworks/dendrites/modules/boosted_dendrites.py]: > I think my wording is confusing. The intuition and math are meant to be identical to the kwinners boosting functionality. It should read "segments that rarely win will be promoted, and the higher the boost strength, the greater this effect". Is that more clear?
